content/blog/2022-03-08-alive/index.md

```diff
@@ -0,0 +1,114 @@
+
+In a similar flavor, undefined behavior allows a compiler to more aggressively optimize code, since while the optimized code must generate the same results whenever the results are defined, it is free to do optimizations and take shortcuts which further modify (mangle?) the code whenever it would produce an undefined result.[^1]
+
+[^1]: If this is unsatisfying, you can also wait until later, where we will talk about this again.
+
+## Main Contributions
+With the background out of the way, I reiterate:
+
+> At its core, Alive is a domain specific language for formally verifying peephole optimizations in LLVM.  Empirically, it found bugs where the optimization could introduce additional undefined behavior in situations where the original code did not.
+
+### Formal Verification of Peephole Optimizations
+A core engine of the paper is in creating a domain specific language which makes it easy to write peephole optimizations which can be formally verified.
+
+#### Basic Overview
+An LLVM instruction can have source preconditions $\phi$ and target postconditions $\bar \phi$, source definedness constraints $\delta$, a source poison-free constraint $\rho$, a source result $\iota$, as well as target definedness, poison-free constraint, and result  $\bar delta$, $\bar \rho$, andt $\bar \iota$ respectively.
+
+*Preconditions* can be defined using a must-analysis for precise constraints, and may-analysis for over/underapproximations. For example, the precise `isPowerOfTwo(%a)` constraint is defined as $p \leftarrow a \neq 0 \wedge a &(a-1) = 0$, while the approximate `mayAlias(%a, %b)` 
+
+*Definedness* simply describes when a variable has a defined output. For example, unsigned division of a/b is defined whenever $b \neq 0$.
+
+*Poison-free* describes when a variable will not generate a poison value, which is a value which could generate undefined behavior under some circumstances. For example, left shift with no sign overflow is poison-free as long as it does not generate a sign overflow, namely as long as $(a << b) >> b = a.
+
+*Result*s are simply their normal meaning – the result of applying the functions in the definition to their inputs.
+
+With this out of the way, we can now understand the somewhat notationally dense definition of correctness in the paper.
+
+<p align="center">
+<img src="paper_snippet.png" alt="A snippet from the paper of section 3.1.2, which describes the correctness criteria" title="Section 3.1.2 Snippet" style="zoom:25%;" />
+</p>
```

Comment:
Maybe a brief, very informal textual summary of what the paper is saying here would help put it into context?

Reply:
Ah, good idea!